### PR TITLE
PIMS-2379 Workflow Project Status Update

### PR DESCRIPTION
**Pull request by Fosol**
## Description

Unit-tests for project status workflow updates.

Fixing validation bugs related to incomplete tasks.

### CODE REVIEW

**Comment by asanchezr** [comment]
Please fix failing backend unit test:

```
[xUnit.net 00:00:10.57]     Pims.Api.Test.Controllers.DisposeControllerTest.SetStatus_WithCode_Success [FAIL]
[xUnit.net 00:00:10.57]       System.NullReferenceException : Object reference not set to an instance of an object.
```

**Comment by Fosol** [comment]
> Please fix failing backend unit test:
> 
> ```
> [xUnit.net 00:00:10.57]     Pims.Api.Test.Controllers.DisposeControllerTest.SetStatus_WithCode_Success [FAIL]
> [xUnit.net 00:00:10.57]       System.NullReferenceException : Object reference not set to an instance of an object.
> ```

fixed

File: backend/dal/Services/Concrete/ProjectService.cs [hunk 1]
```diff
@@ -583,46 +601,14 @@ public void Remove(Project project)
...
-        /// <exception cref="ArgumentNullException">Argument 'status' is required.</exception>
-        /// <exception cref="RowVersionMissingException">Project rowversion is required.</exception>
-        /// <exception cref="KeyNotFoundException">Project does not exist.</exception>
-        /// <exception cref="NotAuthorizedException">User does not have permission to edit project.</exception>
-        /// <exception cref="NotAuthorizedException">User does not have permission to submit request.</exception>
-        /// <exception cref="NotAuthorizedException">User does not have permission to approve request.</exception>
-        /// <exception cref="InvalidOperationException">Invalid project status transition.</exception>
-        /// <exception cref="InvalidOperationException">Denying a project requires a reason to be included in the shared note.</exception>
-        /// <returns></returns>
-        public Project SetStatus(Project project, string workflowCode, int statusId)
```
Comment (Fosol) [inline backend/dal/Services/Concrete/ProjectService.cs:617]: Removing as not needed

File: backend/dal/Services/Concrete/ProjectService.cs [hunk 2]
```diff
@@ -667,17 +652,19 @@ public Project SetStatus(Project project, Workflow workflow, ProjectStatus statu
...
             var fromStatus = this.Context.ProjectStatus
                 .Include(s => s.ToStatus)
+                .Include(s => s.Tasks)
+                .ThenInclude(t => t.Task)
                 .FirstOrDefault(s => s.Id == fromStatusId);
             var toStatus = this.Context.ProjectStatus
                 .Include(s => s.Tasks)
                 .FirstOrDefault(s => s.Id == project.StatusId);
-            if (!fromStatus.ToStatus.Any(s => s.ToStatusId == status.Id) && toStatus.Id != project.StatusId) throw new InvalidOperationException($"Invalid project status transitions from '{fromStatus.Name}' to '{toStatus?.Name}'.");
+            if (!fromStatus.ToStatus.Any(s => s.ToStatusId == project.StatusId)) throw new InvalidOperationException($"Invalid project status transitions from '{fromStatus.Name}' to '{toStatus?.Name}'.");
```
Comment (Fosol) [inline backend/dal/Services/Concrete/ProjectService.cs:661]: Fixing transition validation bug

File: backend/dal/Services/Concrete/ProjectService.cs [hunk 3]
```diff
@@ -667,17 +652,19 @@ public Project SetStatus(Project project, Workflow workflow, ProjectStatus statu
...
 
             // Validate that all required tasks have been completed for the current status before allowing transition from one status to another.
-            var incompleteTaskIds = project.Tasks.Where(t => t.IsCompleted == false).Select(t => t.TaskId);
-            var statusTaskIds = fromStatus.Tasks.Select(t => t.TaskId);
-            var incompleteStatusTaskIds = incompleteTaskIds.Intersect(statusTaskIds);
-            if (originalProject.Tasks.Any(t => !t.Task.IsOptional && !t.IsCompleted && incompleteStatusTaskIds.Contains(t.TaskId))) throw new InvalidOperationException("Not all required tasks have been completed.");
+            var incompleteTaskIds = project.Tasks.Where(t => !t.IsCompleted).Select(t => t.TaskId).Union(originalProject.Tasks.Where(t => !t.IsCompleted).Select(t => t.TaskId)).Distinct();
```
Comment (Fosol) [inline backend/dal/Services/Concrete/ProjectService.cs:664]: Fixing incomplete task validation bug

File: backend/dal/Services/Concrete/ProjectService.cs [hunk 4]
```diff
@@ -726,7 +713,7 @@ public Project SetStatus(Project project, Workflow workflow, ProjectStatus statu
                     break;
                 case ("DE"): // Deny
                     // Must have shared note with a reason.
-                    if (String.IsNullOrWhiteSpace(project.PublicNote)) throw new InvalidOperationException("Shared note must contain a reason before denying project.");
+                    // if (String.IsNullOrWhiteSpace(project.PublicNote)) throw new InvalidOperationException("Shared note must contain a reason before denying project.");
```
Comment (Fosol) [inline backend/dal/Services/Concrete/ProjectService.cs:716]: Not required for this iteration

**Pull request merged**